loki/transform/transform_utilities.py

```diff
@@ -27,10 +27,96 @@
-    'replace_selected_kind'
+    'replace_selected_kind', 'single_variable_declarations', 'single_variable_declaration'
 ]
 
 
+def single_variable_declaration(routine, variables=()):
+    """
+    Modify/extend variable declarations to declare variables specified in ``variables`` in single declarations.
+
+    Parameters
+    ----------
+    routine: :any:`Subroutine`
+        The subroutine in which to modify the variable declarations
+    variables: tuple
+        Variables to grant unique/single declaration for
+    """
+    if variables:
+        decl_map = {}
+        for decl in FindNodes(VariableDeclaration).visit(routine.spec):
+            if len(decl.symbols) > 1:
+                convert = False
+                symbols = []
+                unique_symbols = []
+                for smbl in decl.symbols:
+                    if smbl.name in variables:
+                        convert = True
+                        unique_symbols.append(smbl)
+                    else:
+                        symbols.append(smbl)
+                if convert:
+                    counter = 1
+                    if symbols:
+                        decl_map[decl] = decl.clone(symbols=as_tuple(symbols))
+                    else:
+                        decl_map[decl] = None
+                    for smbl in unique_symbols:
+                        routine.spec.insert(routine.spec.body.index(decl) + counter, (decl.clone(symbols=(smbl,)),))
+                        counter += 1
+        routine.spec = Transformer(decl_map).visit(routine.spec)
+
+
+def single_variable_declarations(routine, strict=True):
+    """
+    Modify/extend variable declarations to only declare one variable each time while preserving the order.
+
+    Parameters
+    ----------
+    routine: :any:`Subroutine`
+        The subroutine in which to modify the variable declarations
+    strict: bool
+        Whether to strictly make unique variable declarations or to only disassemble non-arrays and arrays and among
+        arrays, arrays with differing shapes.
+    """
+    decl_map = {}
+    for decl in FindNodes(VariableDeclaration).visit(routine.spec):
+        if len(decl.symbols) > 1:
+            if strict:
+                counter = 1
+                for i_sdecl, sdecl in enumerate(decl.symbols):
+                    if i_sdecl == 0:
+                        decl_map[decl] = decl.clone(symbols=(sdecl,))
+                    else:
+                        routine.spec.insert(routine.spec.body.index(decl) + counter, (decl.clone(symbols=(sdecl,)),))
+                        counter += 1
+            else:
+                types = [type(smbl) for smbl in decl.symbols]
+                smbls_by_type = {smbl_type: [] for smbl_type in set(types)}
+                for smbl in decl.symbols:
+                    smbls_by_type[type(smbl)].append(smbl)
+                declarations = []
+                for key in smbls_by_type:
```

Comment:
If I'm reading this correctly, I think you're essentially trying to sort the symbols into different groups depending on their shape. May I suggest again a slightly simpler sorting procedure to make this a bit more straightforward? It's untested, so may not work correctly:
```python
symbols_by_shape = defaultdict(list)
for smbl in decl.symbols:
    symbols_by_shape[getattr(smbl, 'shape', None)] += [smbl]
decl_map[decl] = tuple(decl.clone(symbols=as_tuple(smbls)) for smbls in symbols_by_shape.values())
```